File: RadialTree.py
from Tree import Tree, TreeItemDelegate
from Tree import ItemStyle
from Global import Shape

import math
from PySide.QtGui import QPainterPath, QPainter, QPen, QPixmap
from PySide.QtCore import QRect, QPointF


class RadialTree(Tree):
    def __init__(self, parent=None):
        super(RadialTree, self).__init__( parent )
        self._rotateText = False
        self.orderedIndexes = []
        self.rect = QRect( -20, -20, 40, 40 )
        style = ItemStyle()
        style.setShape( Shape.Ellipse )
        
        delegate = TreeItemDelegate( self )
        delegate.setItemStyle( style )
        self.setItemDelegate( delegate )
        
        self.yDistance = 40
        self.xDistance = 40
        
        self.radius = 1
        self.diagonal = 1
        
        self._displayCircle = True
        
    
    def displayCicrcle(self, display = True ):
        self._displayCircle = display
        
        
    def setScrollBarValues(self):
        dw = max( 0, ( self.realSize.width() - self.width() )/2  )
        dh = max ( 0, ( self.realSize.height() - self.height() )/2 )
        self.horizontalScrollBar().setRange( -dw/2, dw/2 )
        self.verticalScrollBar().setRange( -dh, dh )
        self.itemOffset = QPointF( self.width() / 2, self.height() / 2 )
        
        
    def updatePerimeter(self):
        w = self.rect.width() + self.xDistance
        h = self.rect.height() + self.yDistance
        self.diagonal = math.sqrt( w * w + h * h )
        self.perimeter = 0.0
        factor = 1
        for index in self.itemTreePos.keys():
            if self.model().rowCount( index ) == 0:
                factor = self.depth / self.itemTreePos[index].y()
                self.perimeter += self.diagonal * factor
        
        self.radius = self.perimeter / ( 2 * math.pi )
        if self.radius * 2 < self.diagonal * self.depth:
            self.radius = (self.diagonal * self.depth)/2 * 1.5
    
    
    def scan(self, index, left, depth):
        self.orderedIndexes.append( index )
        rows = self.model().rowCount( index )
        if not index.isValid():
            return 0
        elif rows == 0:
            self.setX( index, left )
            self.setY(index, depth + 1)
            return (left + 1, 1)
        
        childDepth = 0
        for r in range( 0, self.model().rowCount( index ) ):
            child = index.child( r, 0 )
            (left, d) = self.scan( child, left, depth + 1 )
            childDepth = max( childDepth, d )
        
        left = self.itemTreePos[ self.model().index(0, index.column(), index) ].x()
        right =  self.itemTreePos[ self.model().index( rows - 1, index.column(), index ) ].x();
        if rows >= 2:
            if rows % 2 == 1:
                r = math.floor( rows/2 ) + 1
                v = self.itemTreePos[ self.model().index( r - 1, index.column(), index ) ].x()
                self.setX( index, v )
            else:
                self.setX( index, float(right + left + 1) / 2.0 )
        else:
            self.setX( index, left )

        self.setY(index, depth )
        return ( right + 1 , childDepth + 1 ) 
        
    
    def positionsInTree(self):
        self.itemTreePos = {}
        self.orderedIndexes = []
        
        (self.left, self.depth) = self.scan( self.model().index(0,0) , 0, 0)
        self.left += 1

        self.positionsInView()
        
        
    def positionsInView(self):
        if len(self.itemTreePos.keys()) == 0:
            return None
        
        self.updatePerimeter()
        
        self.realSize.setWidth( ( self.rect.width() + self.radius ) * 2 )
        self.realSize.setHeight( ( self.rect.height() + self.radius ) * 2 )
        self.setScrollBarValues()
        
        self.itemPos = {}
        
        for index in self.itemTreePos.keys():
            p = self.itemTreePos[ index ]
            self.itemPos[ index ] = QPointF( p.x(), p.y() )
        
        path = QPainterPath()
        path.addEllipse( self.itemOffset + self.rect.center(), self.radius, self.radius )
        
        l = 0
        factor = 0
        for index in self.orderedIndexes:
            if self.model().rowCount( index ) != 0:
                continue
            
            factor = float(self.depth) / float(self.itemPos[index].y())
            
            if self.model().rowCount( index ) == 0:
                l +=  (float(self.diagonal)/2.0) * float(factor)
            
            percent = path.percentAtLength( l )
            self.itemPos[index] = QPointF( percent, self.itemPos[index].y() )

            if self.model().rowCount( index ) == 0:
                l +=  (float(self.diagonal)/2.0) * float(factor)
                       
        self.orderedIndexes.reverse()
        for index in self.orderedIndexes:
            if self.model().rowCount( index ) == 0:
                continue
            left = self.model().index( 0,0, index )
            right = self.model().index( self.model().rowCount(index) - 1,0, index )
            self.itemPos[index] = QPointF( float(self.itemPos[left].x() + self.itemPos[right].x()) / 2.0, self.itemPos[index].y() )        
        
        self.itemPos[self.model().index(0,0)] = QPointF(0, 0)
        
        self.orderedIndexes.reverse()
        
        for index in self.orderedIndexes:
            if self.depth == self.itemPos[index].y():
                factor = 1.0
            else:
                if self.itemPos[index].y() != 0:
                    factor = float(self.depth) / float(self.itemPos[index].y())
                else:
                    factor = 1.0

            radius = float(self.radius) / float(factor)
            circle = QPainterPath()
            circle.addEllipse( self.rect.center(), radius, radius )
            
            self.itemPos[index] = circle.pointAtPercent( self.itemPos[index].x() )
            
        self.itemPos[self.model().index(0,0)] = QPointF(0, 0)

        
    def paintConnections(self, painter, offset):
        if self._displayCircle == True:
            self.paintCircles(painter, offset)

        Tree.paintConnections(self, painter, offset)

        
    def paintCircles(self, painter, offset):
        c = self._connectionPen.color()
        c.setAlpha( c.alpha() * 0.15 )
        pen = QPen(self._connectionPen)
        pen.setColor( c )
        pen.setWidth( pen.width() * 2 )
        painter.setPen( pen )
        painter.setRenderHint( QPainter.Antialiasing )
        index = self.model().index(0,0)
        center = self.itemRect(index).translated( offset.x(), offset.y() ).center()
        for rd in range( 1, int(self.depth) + 1 ):
            r = float(self.radius) * rd / self.depth
            painter.drawEllipse( center, r, r )
            
    def save(self, filename ):
        pix = QPixmap( self.realSize )
        pix.fill( Qt.transparent )
        painter = QPainter( pix )
        painter.setRenderHint( QPainter.Antialiasing )
        painter.drawRect( QRect(QPoint(0,0), self.realSize) )
        x = self.realSize.width() / 2
        y = self.realSize.height() / 2
        self.paintConnections( painter, QPointF( x, y ) )
        self.paintItems(painter , QPointF( x, y ) )
        painter.end()
        return pix.save( filename )

File: Trees/radialtree.cpp
/*
 This file is part of Thistle.
    Thistle is free software: you can redistribute it and/or modify
    it under the terms of the Lesser GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License.
    Thistle is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    Lesser GNU General Public License for more details.
    You should have received a copy of the Lesser GNU General Public License
    along with Thistle.    If not, see <http://www.gnu.org/licenses/>.
 Thistle    Copyright (C) 2013    Dimitry Ernot & Romha Korev
*/
#include "radialtree.h"
#include <QPaintEvent>
#include <QPainter>
#include <QAbstractItemModel>
#include <QFontMetrics>
#include <QScrollBar>
#include <qmath.h>
#include "../kernel/itemdelegate.h"

namespace Thistle {

RadialTree::RadialTree( QWidget* parent ) : AbstractTree( parent ) {
    Q_D( RadialTree );
    d->itemRect = QRect( -20, -20, 40, 40 );
    ItemStyle style = ItemStyle();
    style.setShape( Global::Ellipse );
    d->delegate = new ItemDelegate( this );
    d->delegate->setItemStyle( style );
    setItemDelegate( d->delegate );
    d->yDistance = 40;
    d->xDistance = 40;
    d->itemOffset = QPoint( 0, 0 );
    d->radius = 1;
    d->diagonal = 1;
}

RadialTree::RadialTree( RadialTreePrivate* d, QWidget* parent ) : AbstractTree( d, parent ) {
}


void RadialTree::paintCircles( QPainter& painter, const QPointF& offset ) const {
    const Q_D( RadialTree );
    painter.save();
    QColor c = d->connectionPen.color();
    c.setAlpha( c.alpha() * 0.15 );
    QPen pen( d->connectionPen );
    pen.setColor( c );
    pen.setWidth( pen.width() * 2 );
    painter.setPen( pen );
    painter.setRenderHint( QPainter::Antialiasing );
    QModelIndex index = this->model()->index(0,0);
    QPointF center = itemRect(index).translated( offset.x(), offset.y() ).center();
    for( int rd = 1; rd < d->depth; ++rd ) {
            qreal r = d->radius * rd / d->depth;
            painter.drawEllipse( center, r, r );
    }
    painter.restore();
}


void RadialTree::paintConnections( QPainter& painter, QPointF offset ) const  {
    paintCircles( painter, offset );
    AbstractTree::paintConnections( painter, offset );
}


void RadialTree::positionsInTree() {
    Q_D( RadialTree );
    d->itemTreePos.clear();
    d->orderedIndexes.clear();
    QPointF p = scan( this->model()->index(0,0) , QPointF(0,0) );
    d->left = p.x() + 1;
    d->depth = p.y();
    this->positionsInView();
}


void RadialTree::positionsInView() {
    Q_D( RadialTree );
    if ( d->itemTreePos.isEmpty() || (d->radius == HUGE_VAL) || (d->radius == 0) ) {
            return;
    }
    updatePerimeter();
    if ( (d->radius == HUGE_VAL) || (d->radius == 0) ) {
        return;
    }
    d->realSize.setWidth( d->radius * 2 + d->itemRect.width());
    d->realSize.setHeight( d->radius * 2 + d->itemRect.height() );
    setScrollBarValues();
    d->itemPos = d->itemTreePos;
    QPainterPath path;
    path.addEllipse( d->itemOffset + d->itemRect.center(), d->radius, d->radius );
    qreal l = 0;
    qreal factor = 0;
    Q_FOREACH( QModelIndex index, d->orderedIndexes ) {
        if ( this->model()->rowCount( index ) != 0 ) {
                    continue;
        }
        factor = qreal(d->depth) / qreal(d->itemPos[index].y());
        if ( this->model()->rowCount( index ) == 0 ) {
                l +=	(qreal(d->diagonal)/2.0) * qreal(factor);
        }
        qreal percent = path.percentAtLength( l );
        d->itemPos[index] = QPointF( percent, d->itemPos[index].y() );
        if ( this->model()->rowCount( index ) == 0 ) {
                l +=	(qreal(d->diagonal)/2.0) * qreal(factor);
        }
    }
    QList<QModelIndex> reversedOrderedIndexes;
    Q_FOREACH ( QModelIndex idx, d->orderedIndexes ) {
        reversedOrderedIndexes.prepend( idx );
    }
    Q_FOREACH ( QModelIndex index, reversedOrderedIndexes ) {
            if ( this->model()->rowCount( index ) == 0 ) {
                    continue;
            }
            QModelIndex left = this->model()->index( 0,0, index );
            QModelIndex right = this->model()->index( this->model()->rowCount(index) - 1,0, index );
            d->itemPos[index] = QPointF( float(d->itemPos[left].x() + d->itemPos[right].x()) / 2.0, d->itemPos[index].y() );
    }
    d->itemPos[model()->index(0,0)] = QPointF(0, 0);
    Q_FOREACH ( QModelIndex index, d->orderedIndexes ) {
        if ( d->depth == d->itemPos[index].y() ) {
                factor = 1.0;
        } else {
                if ( d->itemPos[index].y() != 0 ) {
                        factor = qreal(d->depth) / qreal(d->itemPos[index].y());
                } else {
                        factor = 1.0;
                }
        }
        qreal radius = qreal( d->radius ) / qreal(factor);
        QPainterPath circle;
        circle.addEllipse( d->itemRect.center(), radius, radius );
        d->itemPos[index] = circle.pointAtPercent( d->itemPos[index].x() ) + d->itemOffset;
    }
    d->itemPos[ this->model()->index(0,0) ] = d->itemOffset;
}


QPointF RadialTree::scan(QModelIndex index, QPointF leftDepth) {
    Q_D( RadialTree );
    d->orderedIndexes.append( index );
    int rows = this->model()->rowCount( index );
    if ( !index.isValid() ) {
        return QPointF();
    } else if ( rows == 0 ) {
            setX( index, leftDepth.x() );
            setY(index, leftDepth.y() );
            return QPointF(leftDepth.x() + 1, 1);
    }
    qreal childDepth = 0;
    for( int r = 0; r < rows; ++r ) {
            QModelIndex child = index.child( r, 0 );
            QPointF p = scan( child, leftDepth + QPointF( 0, 1 ) );
            leftDepth.setX( p.x() );
            childDepth = qMax( childDepth, p.y() );
    }
    qreal left = d->itemTreePos[ this->model()->index(0, 0, index) ].x();
    qreal right =    d->itemTreePos[ this->model()->index( rows - 1, 0, index ) ].x();
    if ( rows >= 2 ) {
            if ( rows % 2 == 1 ) {
                    qreal r = qFloor( rows/2 ) + 1;
                    qreal v = d->itemTreePos[ this->model()->index( r - 1, 0, index ) ].x();
                    setX( index, v );
            } else {
                    setX( index, ( right + left + 1 ) / 2.0 );
            }
    } else {
            setX( index, left );
    }
    setY(index, leftDepth.y());
    return QPointF( right + 1 , childDepth + 1 );
}


void RadialTree::setRotateText( bool rotate ) {
    Q_D( RadialTree );
    d->rotateText = rotate;
}
/***************************************
**
**                RESOLVING POSITIONS
**
***************************************/


void RadialTree::setScrollBarValues() {
    Q_D( RadialTree );
    QSize s = d->realSize + QSize(50, 50);
    qreal dw = qMax( 0, ( s.width() - width() )/2	);
    qreal dh = qMax ( 0, ( s.height() - height() )/2 );
//    qreal dw = qMax( 0, ( d->realSize.width() - width() )/2 + verticalScrollBar()->width() + 50 );
//    qreal dh = qMax ( 0, ( d->realSize.height() - height() )/2 + horizontalScrollBar()->height() + 50 );
    horizontalScrollBar()->setRange( -dw, dw );
    verticalScrollBar()->setRange( -dh, dh );
    d->itemOffset = QPoint( width() / 2, height() / 2 );
}


void RadialTree::updatePerimeter() {
    Q_D( RadialTree );
    qreal w = d->itemRect.width() + d->xDistance;
    qreal h = d->itemRect.height() + d->yDistance;
    d->diagonal = qSqrt( w * w + h * h );
    d->perimeter = 0;
    qreal factor = 1;
    Q_FOREACH ( QModelIndex index, d->itemTreePos.keys() ) {
        if ( this->model()->rowCount( index ) == 0 ) {
            factor = ( d->depth ) / ( d->itemTreePos.value( index ).y() );
            d->perimeter += d->diagonal * factor;
        }
    }
    d->radius = d->perimeter / ( 2 * M_PI );
    if ( d->radius*2 < d->diagonal * d->depth ) {
        d->radius = ( d->diagonal * d->depth )/2 * 1.5;
    }
}


}